File: DFS&BFS/python/9466.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
T = int(input())
def dfs(n):
  global result
  if visited[n]:
    for i in range(len(route)):
      if route[i] == n: break
      result += 1
    return
  visited[n] = True
  route.append(arr[n])
  dfs(arr[n])


for _ in range(T):
  n = int(input())
  arr = [0] + list(map(int,input().split()))
  visited = [False for _ in range(n+1)]
  result = 0
  for i in range(1,n+1):
    if not visited[i]:
      route = [i]
      dfs(i)
  print(result)

File: DFS&BFS/javascript/9466.js
const input = require('fs').readFileSync('/dev/stdin').toString().split('\n');
const dfs= (n) => {
  if (visited[n]){
    for (let i = 0 ; i<route.length ; i++){
      if (route[i] ===n) break;
      result++;
    }
    return
  }
  visited[n] = true;
  route.push(arr[n]);
  dfs(arr[n]);
}

let arr,n,visited,result;
for (let i = 1 ; i<input.length-1 ; i+=2){
  
  n = Number(input[i]);
  arr = input[i+1].split(' ').reduce((acc,cur)=>{acc.push(Number(cur));return acc},[0]);
  visited = new Array(n+1).fill(false);
  result = 0
  for (let i = 1 ; i<=n ; i++){
    if (!visited[i]) {
      route = [i]
      dfs(i);
    }
  }
  console.log(result)
}